src/AppInstallerCommonCore/GroupPolicy.cpp

```diff
@@ -104,26 +104,110 @@ namespace AppInstaller::Settings
             // Use folding to call each policy validate function.
             (FoldHelper{}, ..., Validate<static_cast<ValuePolicy>(P)>(policiesKey, policies));
         }
+
+        // Reads a list from a Group Policy.
+        // The list is stored in a sub-key of the policies key, and each value in that key is a list item.
+        // Cases not considered by this function because we don't use them:
+        //  - When the list is in an arbitrary key, not a sub key.
+        //  - When the list values are mixed with other values and are identified by a prefix in their names.
+        //  - When the value names are relevant.
+        template<ValuePolicy P>
+        std::optional<typename details::ValuePolicyMapping<P>::value_t> ReadList(const Registry::Key& policiesKey)
+        {
+            using Mapping = details::ValuePolicyMapping<P>;
+
+            auto listKey = policiesKey.SubKey(Mapping::KeyName);
+            if (!listKey.has_value())
+            {
+                return std::nullopt;
+            }
+
+            std::vector<Mapping::item_t> items;
+            for (const auto& value : listKey->Values())
+            {
+                auto item = Mapping::ReadAndValidateItem(value);
+                if (item.has_value())
+                {
+                    items.emplace_back(std::move(item.value()));
+                }
+                else
+                {
+                    AICLI_LOG(Core, Warning, << "Failed to read Group Policy list value. Policy [" << Mapping::KeyName << "], Value [" << value.Name() << ']');
+                }
+            }
+
+            return items;
+        }
+
+        std::optional<SourceFromPolicy> ReadSourceFromRegistryValue(const Registry::Value& item)
+        {
+            auto jsonString = item.TryGetValue<Registry::Value::Type::String>();
+            if (!jsonString.has_value())
+            {
+                AICLI_LOG(Core, Warning, << "Registry value is not a string");
+                return std::nullopt;
+            }
+
+            int stringLength = static_cast<int>(jsonString->length());
+            Json::Value sourceJson;
+            Json::CharReaderBuilder charReaderBuilder;
+            const std::unique_ptr<Json::CharReader> jsonReader(charReaderBuilder.newCharReader());
+            Json::String jsonErrors;
+            if (!jsonReader->parse(jsonString->c_str(), jsonString->c_str() + stringLength, &sourceJson, &jsonErrors))
+            {
+                AICLI_LOG(Core, Warning, << "Registry value does not contain a valid JSON: " << jsonErrors);
+                return std::nullopt;
+            }
+
+            SourceFromPolicy source;
+
+#define READ_SOURCE_ATTRIBUTE(_attr_) \

```

Comment:
(I originally was going to say that you shouldn't use a macro here, but I'm going to just suggest the options)

If we must use a macro here, can we not construct the string 3 times?

But if you want to make it a lambda with default ref capture and use pointer to member syntax that would be cool too.
[https://en.cppreference.com/w/cpp/language/pointer#Pointers_to_data_members](https://en.cppreference.com/w/cpp/language/pointer#Pointers_to_data_members)
```
auto L = [&](const std::string& name, std::string SourceFromPolicy::* member) { ... return true/false; }
if (!L("Name", &SourceFromPolicy::Name)) return {};
```